### PR TITLE
Feat/wasmer wasm alt

### PR DESCRIPTION
**Pull request by ptitSeb**
# Description
Abiility to build wasmer compile engine as WebAssembly, using a new CLI program `wasmer-compiler-cli`.
This is done with following steps:
 * New wasmer-engine-universal-artifact (extracted from wasmer-engine-universal), that handle only compilation and serialization.
 * wasmer-engine-universal will uses this crate to create the artifact binary.
 * wasmer-compiler-cli uses this crate for offline compilation.

### CODE REVIEW

**Review by epilys — APPROVED** [review]

File: lib/cli-compiler/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+[[bin]]
+name = "wasmer-compiler"
+path = "src/bin/wasmer_compiler.rs"
+doc = false
+
+[dependencies]
+wasmer-engine-universal-artifact = { version = "=2.2.1", path = "../universal-artifact", features = ["compiler"] }
+wasmer-compiler = { version = "=2.2.1", path = "../compiler" }
+wasmer-types = { version = "=2.2.1", path = "../types" }
+wasmer-vfs  = { version = "=2.2.1", path = "../vfs", default-features = false, features = ["host-fs"] }
```
Comment (Amanieu) [inline lib/cli-compiler/Cargo.toml:24]: Why does wasmer-compiler-cli depend on the VFS? That's only used when running an instance, not compiling it.

File: lib/cli/src/bin/wasmer_compiler.rs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+use wasmer_cli::cli::wasmer_main;
```
Comment (Amanieu) [inline lib/cli/src/bin/wasmer_compiler.rs:1]: This file is not needed any more.

File: lib/cli/Cargo.toml [hunk 3]
```diff
@@ -24,19 +24,12 @@ doc = false
 required-features = ["headless"]
```
Comment (Amanieu) [inline lib/cli/Cargo.toml:24]: The changes in this file are not needed any more.

File: lib/vfs/src/host_fs.rs [hunk 4]
```diff
@@ -9,6 +9,8 @@ use std::fs;
 use std::io::{self, Read, Seek, Write};
 #[cfg(unix)]
 use std::os::unix::io::{AsRawFd, RawFd};
+#[cfg(target_arch = "wasm32")]
```
Comment (Amanieu) [inline lib/vfs/src/host_fs.rs:12]: These vfs changes shouldn't be needed any more now that wasmer-compiler-cli doesn't depend on vfs.

File: lib/compiler-singlepass/Cargo.toml [hunk 5]
```diff
@@ -34,6 +35,7 @@ maintenance = { status = "actively-developed" }
 
 [features]
 default = ["std", "rayon", "unwind", "avx"]
+wasm = ["std", "avx"]
```
[the remark this answers is not in the document]
Comment (ptitSeb) [reply lib/compiler-singlepass/Cargo.toml:38]: unwind is a runtime concept, and wasm build only handle compile, nothing about runtime.

File: lib/vm/Cargo.toml [hunk 6]
```diff
@@ -24,10 +24,13 @@ serde = { version = "1.0", features = ["derive", "rc"] }
 rkyv = { version = "0.7.20", optional = true }
 loupe = { version = "0.1", features = ["enable-indexmap"] }
 enum-iterator = "0.7.0"
-corosensei = { version = "0.1.2" }
 scopeguard = "1.1.0"
 lazy_static = "1.4.0"
 
+[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
+region = { version = "3.0" }
+corosensei = { version = "0.1.2" }
```
[the remark this answers is not in the document]
Comment (ptitSeb) [reply lib/vm/Cargo.toml:30]: Right. I have removed the conditionnal part.

File: lib/compiler-singlepass/Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ptitSeb) [reply lib/compiler-singlepass/Cargo.toml]: my bad, it build fine. I have added it to wasm build.

File: lib/types/src/trapcode.rs [hunk 7]
```diff
@@ -162,3 +166,72 @@ mod tests {
         assert_eq!("users".parse::<TrapCode>(), Err(()));
     }
 }
+
+/// Stores trace message with backtrace.
+#[derive(Debug)]
+pub enum Trap {
```
[the remark this answers is not in the document]
Comment (ptitSeb) [reply lib/types/src/trapcode.rs]: Ok, moved back to VM.

File: lib/types/src/memory.rs [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+use loupe::MemoryUsage;
+#[cfg(feature = "enable-rkyv")]
+use rkyv::{Archive, Deserialize as RkyvDeserialize, Serialize as RkyvSerialize};
+#[cfg(feature = "enable-serde")]
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+
+/// Error type describing things that can go wrong when operating on Wasm Memories.
+#[derive(Error, Debug, Clone, PartialEq, Hash)]
+pub enum MemoryError {
```
[the remark this answers is not in the document]
Comment (ptitSeb) [reply lib/types/src/memory.rs]: ok, done.

File: lib/cli-compiler/Cargo.toml (hunk 1 above)
Comment (ptitSeb) [reply lib/cli-compiler/Cargo.toml:24]: no specific reason, it's a leftover, I'll remove it.

File: lib/cli/src/bin/wasmer_compiler.rs (hunk 2 above)
Comment (ptitSeb) [reply lib/cli/src/bin/wasmer_compiler.rs:1]: yep

**Comment by Amanieu** [comment]
Also can you rebase the branch on top of the latest master? There are a few unrelated commits in there.

File: lib/cli/Cargo.toml (hunk 3 above)
Comment (ptitSeb) [reply lib/cli/Cargo.toml]: ok

File: lib/vfs/src/host_fs.rs (hunk 4 above)
Comment (ptitSeb) [reply lib/vfs/src/host_fs.rs]: ok, done.

**Comment by syrusakbary** [comment]
@Amanieu @epilys should we merge this into the `wasmer3` branch instead of `master` ?

**Comment by Amanieu** [comment]
I don't think there are any API breaking changes, so it's not necessary.

**Comment by Amanieu** [comment]
Also could you rebase on top of master (`git fetch && git rebase origin/master && git push --force`) instead of merging?

**Comment by ptitSeb** [comment]
> Also could you rebase on top of master (`git fetch && git rebase origin/master && git push --force`) instead of merging?

I alway have a conflict with the Update to wasmer 0.83:
```
Auto-merging lib/compiler-cranelift/src/translator/code_translator.rs
CONFLICT (content): Merge conflict in lib/compiler-cranelift/src/translator/code_translator.rs
Auto-merging lib/compiler-singlepass/src/codegen.rs
error: could not apply fc461f605... Update wasmparser to 0.83
```

**Comment by ptitSeb** [comment]
Finaly got it. In needed to use interactive rebase to remove the strange merge I created earlier and that was incorrect. Seems clean now.

File: lib/api/src/js/mod.rs [hunk 9]
```diff
@@ -41,7 +41,6 @@ mod resolver;
 mod store;
 mod trap;
 mod types;
-mod utils;
```
[the remark this answers is not in the document]
Comment (Amanieu) [reply lib/api/src/js/mod.rs:44]: This is not resolved: the file is still not deleted.

[the remark this answers is not in the document]
Comment (ptitSeb) [reply lib/api/src/js/mod.rs:44]: what file? `lib/api/src/js/utils.rs`?  It's not there.
image

[the remark this answers is not in the document]
Comment (Amanieu) [reply lib/api/src/js/mod.rs:44]: Huh the github "Files changed" didn't show it as deleted.

**Pull request merged**